godot-macros/src/derive_from_variant.rs

```diff
@@ -0,0 +1,166 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::util::decl_get_info;
+use crate::ParseResult;
+use proc_macro2::TokenStream;
+use quote::quote;
+use venial::Declaration;
+use venial::StructFields;
+
+pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+    let (where_, generic_params, ident, string_ident) = decl_get_info(&decl);
+    let mut body = quote! {
+        let root = variant.try_to::<godot::builtin::Dictionary>()?;
+        let root = root.get(#string_ident).ok_or(godot::builtin::VariantConversionError::BadType)?;
+    };
+    match decl {
+        Declaration::Struct(s) => match s.fields {
+            venial::StructFields::Unit => {
+                body = quote! {
+                    #body return Ok(Self);
+                }
+            }
+            venial::StructFields::Tuple(fields) => {
+                if fields.fields.len() == 1 {
+                    body = quote! {
+                        #body
+                        let root = root.try_to()?;
+                        Ok(Self(root))
+                    };
+                } else {
+                    let mut app = quote! {};
+                    for _ in fields.fields.iter() {
+                        app = quote! {
+                            #app root.pop_front().ok_or(godot::builtin::VariantConversionError::MissingValue)?.try_to()?,
+                        }
+                    }
+                    body = quote! {
+                        #body
+                        let mut root = root.try_to::<godot::builtin::Array<godot::builtin::Variant>>()?;
+                        Ok(Self(#app))
+                    };
+                }
+            }
+            venial::StructFields::Named(fields) => {
+                let mut app = quote! {};
+                for (field, _) in fields.fields.iter() {
+                    let ident = &field.name;
+                    let string_ident = &field.name.to_string();
+                    app = quote! {
+                        #app #ident : root.get(#string_ident).ok_or(godot::builtin::VariantConversionError::MissingValue)?.try_to()?,
```

Comment:
In the future, I maybe would like to add support for writing things like 
```rust
expr.ok_or(godot::builtin::VariantConversionError::MissingValue).with_context("Failed to find a value for #ident while trying to blablabla")?.try_to()?;
```

Where whith_context maps the error, returns the same type of error but improve error messages.

But it doesn't seem incompatible with try_convert so I'll add it (not sure it will make a big compile time difference tho but maybe for big projects).

Reply:
Problem: with your previous comment about not returning Ok(...) unconditionally I refactored most things into intermediate variables which separates ok_or and try_to calls...
I could unite them again if you think it's important !

Reply:
> In the future, I maybe would like to add support for writing things like ...

But how would that contradict a separate function? You could still have `try_convert(expr, context)?` or something.